### DISCUSSION TITLE
How do you find the WMI calls?

### DISCUSSION THREAD

**Discussion by flukejones** [seerge/g-helper#532] [General]
Hi!

I am the maintainer of https://gitlab.com/asus-linux/asusctl, and also write a lot of the Linux kernel patches to expose WMI methods to users. Most of the time I find WMI methods by trial and error, by searching others repos, or by asking ASUS.

Do you @seerge use any particular method to find these? Reverse engineering drivers? I see you have a few in your project I haven't yet seen before.

**Comment by seerge** [comment]
@flukejones hello! Nice to have you here, I use your **asusctl** on my dual boot ubuntu :) 

I mainly use DEVS / DSTS  methods for the app, and I call them via ACPI proxy-layer that asus has created, as then it's possible to call them even without admin privileges at all.

Main unknown here is "device ID" param that is being used by those methods. Many of them (like ones for fan curves / or GPU disabling) I took from linux kernel sources, discovered by people like you :)

As for the rest - it's also trial and error.

1. I did a scan script (actually did it as way faster exe as well, but it get's flagged by windows defender so i don't post it) that goes over all possible device IDS and runs DSTS method on them, trying to "read" value. Often if device exists for DSTS (at least responds something reasonable) - then it will exist for DEVS too. This way all power limits were discovered. There is a topic dedicated to that https://github.com/seerge/g-helper/issues/38 

4. It's also possible to do an ACPI dump using method explained here https://github.com/seerge/g-helper/issues/38#issuecomment-1502133557 then you can get some insights and try to guess what it does.

**Reply by flukejones** [reply]
Oh nice resources, thank you. I've just created a bunch of kernel patches based on some of the acpi methods you've got listed in your sources - and I guess the big feature will be the gpu and cpu boost (a lot of people been asking for this!). I suppose the big question I have for you is how do you determine what something does if you don't have a laptop with that method?

I tend to dump a lot of stuff in a disorganised repo, you might find something interesting in it? It's good for grepping through dts files at least https://gitlab.com/asus-linux/reverse-engineering

FYI I am in contact with asus engineers under NDA, I can get some limited info from them.

**Reply by seerge** [reply]
@flukejones for CPU boost it was a pure logic. I knew that Armoury somehow sets CPU / APU / GPU limits, and most probably does this via it's own ACPI interface. So I just looked for something like that and when you see a list of a very similar endpoints coming in a row, my guess would be they do something.

Then I was just setting values to each endpoint, and running benchmarks, looking what it does. on my G14 2022 and on G16 2023 (that I "borrowed") for that purpose. G16 comes with Nvidia GPU (in contrary to G14 with AMD) and it exposed C0 / C2 endpoints that didn't exist before. So my guess was they are related to GPU sliders in Amoury.

```
0x001200A0 - SPL almost everywhere ( Total Platform limit  on all AMD platforms )
0x001200A1 - EDC_LIMIT_VDD
0x001200A2 - TDC_LIMIT_VDD
0x001200A3 - sPPT

0x001200B0 - CPU limit on all AMD platforms
0x001200B1 - Total Platform limit on all AMD platforms

0x001200C0 - does nothing on G14 2022 (NVIDIA GPU dynamic boost 5 ... 20+W)
0x001200C1 - fPPT
0x001200C2 - does nothing on G14 2022 (NVIDIA GPU temp limit 75 ... 87 C)
```

**Reply by seerge** [reply]
@flukejones i'm now looking for an endpoint that is responsible for sRGB display switch ... mb asus knows? :D

**Reply by flukejones** [reply]
I can definitely ask. What is the feature labelled as in Windows, and is it available in Armory Crate?

**Reply by seerge** [reply]
It's this one https://github.com/seerge/g-helper/issues/520 , literally possibility to switch game visuals (i.e. display color modes), but mainly sRGB. This is one of the last features that is available in AC and not in G-helper ))

**Reply by seerge** [reply]
@flukejones https://github.com/seerge/g-helper/issues/405 - live example how to figure out how some things works on devices that I don't own :)))

**Reply by flukejones** [reply]
Thanks! I'm unsure I can translate that to linux kernel. Currently adding the tunables.

@seerge how did you determine the min/max for each CPU/GPU tunable? And defaults?

**Reply by seerge** [reply]
@flukejones experimentally, i don't know default values for CPU tho (As they are different per model), for GPU it's only max temp (87) and max boost (20 or 25w)

**Reply by flukejones** [reply]
I checked with Armory Crate too.

Do you know which precise WMI methods match PL1 and PL2?

**Reply by seerge** [reply]
A0, A3 and C1 i think (same as for AMD)

**Reply by flukejones** [reply]
Alright so for:
```
    public const int PPT_TotalA0 = 0x001200A0;  // SPL (Total limit for all-AMD models)
    public const int PPT_EDCA1 = 0x001200A1;  // CPU EDC
    public const int PPT_TDCA2 = 0x001200A2;  // CPU TDC
    public const int PPT_APUA3 = 0x001200A3;  // sPPT (long boost limit)
```
I have `0x001200A0` and `0x001200A3`, and on a GU603Z, so Intel + Nvidia. I believe the limits are:
- `0x001200A0` = 60w
- `0x001200A3` = 135w
So that would make `A0` the `PL1` setting and `A3` the `PL2` setting. So what would `C1` be?

It's important to me to get the naming and function label correct since once I get these exposed in Linux it's pretty much set in concrete. So for the initial patch I worked off the labels you have, and I'm not sure `PPT_APUA3` is correct?

Sorry to be a bother about it, but as I say, once it's in the Linux kernel I can't change it.

For example I have:
```c
/* Tunables for AUS ROG laptops */
#define ASUS_WMI_DEVID_PPT_TOTAL	0x001200A0 // PL1? Base TDP?
#define ASUS_WMI_DEVID_PPT_EDC		0x001200A1
#define ASUS_WMI_DEVID_PPT_TDC		0x001200A2
#define ASUS_WMI_DEVID_PPT_APU_SPPT	0x001200A3 // needs better label, not APU, maybe should be PL2?
#define ASUS_WMI_DEVID_PPT_APU_LIMIT	0x001200B0
#define ASUS_WMI_DEVID_PPT_LIMIT_SLOW	0x001200B1
#define ASUS_WMI_DEVID_PPT_APU_FPPT	0x001200C1
#define ASUS_WMI_DEVID_NV_GPU_BOOST	0x001200C0
#define ASUS_WMI_DEVID_NV_GPU_TEMP_TARGET	0x001200C2
```

**Reply by seerge** [reply]
@flukejones It could be that C1 does nothing, i'm not very familiar with intel limits, but as far as i understand pl1 - is a slow boost and pl2 is a fast boost ? 

The best way to test is to literally set some distinctive values and then load everything with cinebench23 and watch CPU power consumption.

**Reply by flukejones** [reply]
Additionally it looks like the default boot values for PL1 and PL2 are the actual max values, and these CPU tunables are meant as a tweak-downwards thing.

**Reply by seerge** [reply]
Or it could be that A0 is TDP, A1 is PL1 and C1 is PL2 ...

**Reply by seerge** [reply]
Usually default values are different per mode (0x120075), for Trubo (1) they are highest, for Silent (2) lowest

**Reply by flukejones** [reply]
> Usually default values are different per mode (0x120075), for Trubo (1) they are highest, for Silent (2) lowest

I guess you mean `0x00120075`. Do you know if the actual package TDp changes with that? We've had `0x00120075` exposed in kernel for a while now but it always seemed more like it only controlled the fans, and since the fans were changing, the effect thermal limit changed due to the CPU throttling from heat. But I'm not sure now.

**Reply by seerge** [reply]
@flukejones on many models (incl G14) it definitely does change power limits as well, so it's not only fans.

Read it as "preset modes" in bios. That you can later customize by overriding fan curves and PPTs

**Reply by flukejones** [reply]
@seerge we can't get what the PL1 and PL2 was changed to can we?

I've tried a few methods, and the package limits never change - those are hard set at factory. Plus the WMI methods give fuck all feedback, and there doesn't appear to be anything in dts to give indication of limits or defaults.

When I switch profiles with `0x00120075` the following never changes:
```
cpu0: PKG Limit #1: ENabled (200.000 Watts, 56.000000 sec, clamp DISabled)
cpu0: PKG Limit #2: DISabled (135.000 Watts, 0.002441* sec, clamp DISabled)
```
(using `sudo turbostat --Summary --interval 5 --show Avg_MHz,Busy%,Bzy_MHz,IRQ,PkgTmp,PkgWatt,GFXWatt`).

That at least gives the PL2, which provides the maximum value for us. But the rest of the MSR doesn't give much.

**Reply by seerge** [reply]
Try to set 0x00120075 to 4, it's a "custom" mode and some devices apparently require that to allow PPT limit changes

**Reply by flukejones** [reply]
I might not have been clear - the effects are there - you can tweak downwards from the known maximum. But the actual PL1 and PL2 the CPU is hard set at never changes (even though they have changed in acpi, and the effect is measurable).

The end result is that although it works there is no way to verify the setting beyond caching the input, and there is no way to easily find the maximum beyond `turbostat`, and even then that doesn't report the correct PL1.

**Reply by seerge** [reply]
@flukejones ah, that's indeed seem to be the case with intel CPUs and way asus does this (this topic confirms that https://github.com/seerge/g-helper/issues/405 ) , so actual PL1/PL2 values don't change if you would query them, but limit is there. I assume this is same as it happens if you would do that via Armoury Crate.

On AMD CPUs it's a bit easier, and at least values change. 

From my observations limits always work downwards, and sometimes a bit upwards (up to 10%) from defaults, depending on model. On g14 for example, default CPU limit would be 80W (in turbo mode), but you can rise it up to 90W.

**Comment by flukejones** [comment]
I see you haven't used:
```csharp
    public const int PPT_EDCA1 = 0x001200A1;  // CPU EDC
    public const int PPT_TDCA2 = 0x001200A2;  // CPU TDC
```

What do we know about these?

**Reply by seerge** [reply]
@flukejones it's max [EDC and TDC](https://linustechtips.com/topic/1184111-what-do-tdc-and-edc-mean-in-ryzen-master/) values, they are not of use for an average user anyway

**Reply by flukejones** [reply]
Ah thanks. I think I'll skip exposing those maybe. Not sure.
